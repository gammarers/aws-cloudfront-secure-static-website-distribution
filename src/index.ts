import * as crypto from 'crypto';
import * as cdk from 'aws-cdk-lib';
import * as acm from 'aws-cdk-lib/aws-certificatemanager';
import * as cloudfront from 'aws-cdk-lib/aws-cloudfront';
import * as origins from 'aws-cdk-lib/aws-cloudfront-origins';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

export interface CloudFrontSecureStaticWebsiteDistributionProps {
  readonly comment?: string;
  readonly customDomain: string;
  readonly customDomainCertificate: acm.ICertificate;
  readonly originBucket: s3.IBucket;
  readonly priceClass?: cloudfront.PriceClass;
  readonly loggingBucket?: s3.IBucket;
}

export class CloudFrontSecureStaticWebsiteDistribution extends cloudfront.Distribution {

  constructor(scope: Construct, id: string, props: CloudFrontSecureStaticWebsiteDistributionProps) {

    // ðŸ‘‡ Create random 8 length string
    const random: string = crypto.createHash('shake256', { outputLength: 4 })
      .update(`${cdk.Names.uniqueId(scope)}.${props.customDomain}`)
      .digest('hex');

    super(scope, id, {
      enabled: true,
      comment: props.comment ?? 'Static Website distribution',
      defaultRootObject: 'index.html',
      certificate: props.customDomainCertificate,
      domainNames: [
        props.customDomain,
      ],
      minimumProtocolVersion: cloudfront.SecurityPolicyProtocol.TLS_V1_2_2021,
      sslSupportMethod: cloudfront.SSLMethod.SNI,
      httpVersion: cloudfront.HttpVersion.HTTP2_AND_3,
      // webAclId: props.wafAclId,
      enableLogging: !!props.loggingBucket,
      logBucket: props.loggingBucket,
      logFilePrefix: props.customDomain + '/original-log/',
      defaultBehavior: {
        allowedMethods: cloudfront.AllowedMethods.ALLOW_GET_HEAD_OPTIONS,
        cachedMethods: cloudfront.CachedMethods.CACHE_GET_HEAD_OPTIONS,
        cachePolicy: cloudfront.CachePolicy.CACHING_OPTIMIZED,
        compress: true,
        viewerProtocolPolicy: cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
        origin: origins.S3BucketOrigin.withOriginAccessControl(props.originBucket, {
          originAccessControl: new cloudfront.S3OriginAccessControl(scope, 'idxxxx', {
            originAccessControlName: `secure-static-website-${random}-origin-access-controle`,
            description: 'secure static website orgin access controle',
            signing: cloudfront.Signing.SIGV4_ALWAYS,
          }),
          originAccessLevels: [
            cloudfront.AccessLevel.READ,
          ],
        }),
        responseHeadersPolicy: new cloudfront.ResponseHeadersPolicy(scope, 'ResponseHeadersPolicy', {
          responseHeadersPolicyName: `secure-static-website-response-header-${random}-policy`,
          comment: 'A default policy',
          securityHeadersBehavior: {
            //contentSecurityPolicy: { contentSecurityPolicy: 'default-src https:;', override: true },
            contentTypeOptions: { override: true },
            frameOptions: {
              frameOption: cloudfront.HeadersFrameOption.SAMEORIGIN,
              override: true,
            },
            referrerPolicy: {
              referrerPolicy: cloudfront.HeadersReferrerPolicy.STRICT_ORIGIN_WHEN_CROSS_ORIGIN,
              override: true,
            },
            strictTransportSecurity: {
              accessControlMaxAge: cdk.Duration.seconds(31536000), // 1 year
              preload: true,
              includeSubdomains: true,
              override: true,
            },
            xssProtection: {
              protection: true,
              modeBlock: true,
              override: true,
            },
          },
          customHeadersBehavior: {
            customHeaders: [
              {
                header: 'Cache-Control',
                value: 'no-cache, no-store, private',
                override: true,
              },
              {
                header: 'pragma',
                value: 'no-cache',
                override: true,
              },
            ],
          },
        }),
      },
      errorResponses: [
        {
          ttl: cdk.Duration.seconds(300),
          httpStatus: 403,
          responseHttpStatus: 404,
          responsePagePath: '/404.html',
        },
        {
          ttl: cdk.Duration.seconds(300),
          httpStatus: 404,
          responseHttpStatus: 404,
          responsePagePath: '/404.html',
        },
      ],
      priceClass: props.priceClass,
    });
  }
}