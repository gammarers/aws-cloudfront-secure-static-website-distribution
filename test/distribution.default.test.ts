import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as acm from 'aws-cdk-lib/aws-certificatemanager';
import * as cloudfront from 'aws-cdk-lib/aws-cloudfront';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { CloudFrontSecureStaticWebsiteDistribution } from '../src';

describe('CustomResource Testing', () => {

  const app = new App();
  const stack = new Stack(app, 'TestingStack', {
    env: {
      account: '123456789012',
      region: 'us-east-1',
    },
  });

  const originBucket = new s3.Bucket(stack, 'StaticWebsiteOriginBucket', {
    encryption: s3.BucketEncryption.S3_MANAGED,
    versioned: false,
  });

  const certificate = new acm.Certificate(stack, 'StaticWebsiteCertificate', {
    domainName: 'static.example.com',
    // validation: acm.CertificateValidation.fromDns(hostedZone),
  });

  const distribution = new CloudFrontSecureStaticWebsiteDistribution(stack, 'CloudFrontSecureStaticWebsiteDistribution', {
    customDomain: 'static.example.com',
    customDomainCertificate: certificate,
    originBucket: originBucket,
  });

  const template = Template.fromStack(stack);

  it('Is GitLab Self Managed Connectrion Host CustomResource', () => {
    expect(distribution).toBeInstanceOf(cloudfront.Distribution);
  });

  it('Should match snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });

});
