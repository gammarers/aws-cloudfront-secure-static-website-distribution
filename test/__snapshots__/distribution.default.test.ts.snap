// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomResource Testing Should match snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CloudFrontSecureStaticWebsiteDistributionA6621912": {
      "Properties": {
        "DistributionConfig": {
          "Aliases": [
            "static.example.com",
          ],
          "Comment": "Static Website distribution",
          "CustomErrorResponses": [
            {
              "ErrorCachingMinTTL": 300,
              "ErrorCode": 403,
              "ResponseCode": 404,
              "ResponsePagePath": "/404.html",
            },
            {
              "ErrorCachingMinTTL": 300,
              "ErrorCode": 404,
              "ResponseCode": 404,
              "ResponsePagePath": "/404.html",
            },
          ],
          "DefaultCacheBehavior": {
            "AllowedMethods": [
              "GET",
              "HEAD",
              "OPTIONS",
            ],
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "CachedMethods": [
              "GET",
              "HEAD",
              "OPTIONS",
            ],
            "Compress": true,
            "ResponseHeadersPolicyId": {
              "Ref": "ResponseHeadersPolicy13DBF9E0",
            },
            "TargetOriginId": "TestingStackCloudFrontSecureStaticWebsiteDistributionOrigin1DF30DFF4",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2and3",
          "IPV6Enabled": true,
          "Origins": [
            {
              "DomainName": {
                "Fn::GetAtt": [
                  "StaticWebsiteOriginBucket6EE753C0",
                  "RegionalDomainName",
                ],
              },
              "Id": "TestingStackCloudFrontSecureStaticWebsiteDistributionOrigin1DF30DFF4",
              "OriginAccessControlId": {
                "Fn::GetAtt": [
                  "idxxxxAFC70BC3",
                  "Id",
                ],
              },
              "S3OriginConfig": {
                "OriginAccessIdentity": "",
              },
            },
          ],
          "ViewerCertificate": {
            "AcmCertificateArn": {
              "Ref": "StaticWebsiteCertificate963AEB98",
            },
            "MinimumProtocolVersion": "TLSv1.2_2021",
            "SslSupportMethod": "sni-only",
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "ResponseHeadersPolicy13DBF9E0": {
      "Properties": {
        "ResponseHeadersPolicyConfig": {
          "Comment": "A default policy",
          "CustomHeadersConfig": {
            "Items": [
              {
                "Header": "Cache-Control",
                "Override": true,
                "Value": "no-cache, no-store, private",
              },
              {
                "Header": "pragma",
                "Override": true,
                "Value": "no-cache",
              },
            ],
          },
          "Name": "secure-static-website-response-header-a9c37bba-policy",
          "SecurityHeadersConfig": {
            "ContentTypeOptions": {
              "Override": true,
            },
            "FrameOptions": {
              "FrameOption": "SAMEORIGIN",
              "Override": true,
            },
            "ReferrerPolicy": {
              "Override": true,
              "ReferrerPolicy": "strict-origin-when-cross-origin",
            },
            "StrictTransportSecurity": {
              "AccessControlMaxAgeSec": 31536000,
              "IncludeSubdomains": true,
              "Override": true,
              "Preload": true,
            },
            "XSSProtection": {
              "ModeBlock": true,
              "Override": true,
              "Protection": true,
            },
          },
        },
      },
      "Type": "AWS::CloudFront::ResponseHeadersPolicy",
    },
    "StaticWebsiteCertificate963AEB98": {
      "Properties": {
        "DomainName": "static.example.com",
        "DomainValidationOptions": [
          {
            "DomainName": "static.example.com",
            "ValidationDomain": "example.com",
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestingStack/StaticWebsiteCertificate",
          },
        ],
        "ValidationMethod": "EMAIL",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
    "StaticWebsiteOriginBucket6EE753C0": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "StaticWebsiteOriginBucketPolicy83D2F3E9": {
      "Properties": {
        "Bucket": {
          "Ref": "StaticWebsiteOriginBucket6EE753C0",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:GetObject",
              "Condition": {
                "StringEquals": {
                  "AWS:SourceArn": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":cloudfront::",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":distribution/",
                        {
                          "Ref": "CloudFrontSecureStaticWebsiteDistributionA6621912",
                        },
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "cloudfront.amazonaws.com",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "StaticWebsiteOriginBucket6EE753C0",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "idxxxxAFC70BC3": {
      "Properties": {
        "OriginAccessControlConfig": {
          "Description": "secure static website orgin access controle",
          "Name": "secure-static-website-a9c37bba-origin-access-controle",
          "OriginAccessControlOriginType": "s3",
          "SigningBehavior": "always",
          "SigningProtocol": "sigv4",
        },
      },
      "Type": "AWS::CloudFront::OriginAccessControl",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
